Initial support for ARM Cortex-R

bbolen opened the pull request:
There have been a couple of requests for the addition of ARM Cortex-R support, e.g. https://github.com/zephyrproject-rtos/zephyr/issues/6545 and https://github.com/zephyrproject-rtos/zephyr/issues/5540.

I have started work for Cortex-R4 support for a project my company is doing.  This PR represents work that gets the "shell" sample working on an R4.  Basic context switching and interrupts using an R4 connected to a GIC400 are working.

The work and testing were done on a custom SoC so the PR does not include the SoC code since it won't be very useful to anyone.

Closes #6545
Closes #5540



Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
interrupt_controller: gic: Add support for the GIC400

The GIC400 is a common interrupt controller that can be used with the
Cortex A and R series processors.  This patch adds basic interrupt
handling for the GIC, but does not handle multiple routing or
priorities.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: Add Cortex-R support

This adds initial Cortex-R support for interrupts and context switching.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

ulfalizer commented on arch/arm/core/Kconfig:
```diff
@@ -26,6 +26,19 @@ config CPU_CORTEX_M
+config CPU_CORTEX_R
+	bool
+	# Omit prompt to signify "hidden" option
+	default n
```
`n` is the implicit default value for `bool` symbols, so could remove the default.

That generates clearer documentation at http://docs.zephyrproject.org/reference/kconfig/index.html as well ("Implicitly defaults to `n`" instead of "default n if \<possibly complicated dependencies>").

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+	  implemented as big endian.
+
+config STACK_ALIGN_DOUBLE_WORD
+	bool
```
Can be shortened to `bool "Align stacks..."`

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+
+config RUNTIME_NMI
+	bool
+	prompt "Attach an NMI handler at runtime"
```
Ditto here.

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+
+config FAULT_DUMP
+	int
+	prompt "Fault dump level"
```
Ditto here (`int "Fault dump level"`).

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+
+config LDREX_STREX_AVAILABLE
+	bool
+	default y
```
Not super important, but could use `def_bool y` as well (and ditto for the other `bool` + `default y`s).

ulfalizer commented on arch/arm/core/Kconfig:
```diff
@@ -26,6 +26,19 @@ config CPU_CORTEX_M
 
+config CPU_CORTEX_R
+	bool
+	# Omit prompt to signify "hidden" option
```
I wonder how helpful these comments are in practice. There's a gazillion of them in Zephyr.

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+config ISA_THUMB2
+	bool
+	# Omit prompt to signify "hidden" option
+	default n
```
Redundant `default n`.

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+
+config ISA_ARM
+	bool
+	default n
```
Redundant `default n`.

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+config ARMV7_R_ARMV8_R_MAINLINE
+	bool
+	# Omit prompt to signify "hidden" option
+	default n
```
Redundant `default n`.

ulfalizer commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,161 @@
+
+config XIP
+	default y
+endmenu
```
Could put a blank line before `endmenu`.

ulfalizer reviewed (changes requested):
Just some Kconfig nits.

galak commented:
@bbolen have you tried to see if you can port qemu and the xlnx-zynqmp platform?

bbolen commented:
@ulfalizer Thanks for the comments.  Those should all be fixed now.

bbolen commented:
@galak I saw that yesterday trying to find a way for others to replicate my work.  I will keep looking at it, but I can't guarantee when I may have something useful with respect to qemu.

ulfalizer reviewed (approved):
Kconfig parts look good to me.

pizi-nordic commented on include/arch/arm/cortex_r/asm_inline_gcc.h:
```diff
@@ -0,0 +1,158 @@
+/* ARM Cortex-M GCC specific public inline assembler functions and macros */
```
IMHO with some modifications in _arch_irq_lock()/_arch_irq_unlock() this file could be common for Cortex-M and R.

ioannisg reviewed (changes requested):
Went through most of the /arch code; added a few comments about re-using code and about Kconfig options.

bbolen commented on drivers/interrupt_controller/gic.c:
```diff
@@ -0,0 +1,230 @@
+	 * Set priority on PPI and SGI interrupts
+	 */
+	for (i = 0; i < 32; i += 4)
+		sys_write32(0xa0a0a0a0, GICD_IPRIORITYRn + i);
```
I see two possible paths to get the flags and priority from gic.c.  1) add an intr_set_priority() callback to the irq_next_level_api struct and call it from arch/arm/core/irq_manage.c:_irq_priority_set() OR 2) redefine _ARCH_IRQ_CONNECT for Cortex-R to pass flags_p instead of 0 to _ISR_DECLARE and get to the flags through the isr_sw_table.

Do you have a preference between these two, or another option?

bbolen commented on arch/arm/core/cortex_r/irq_init.c:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+void _IntLibInit(void)
```
It should be empty.  I can move this to arch/arm/include/kernel_arch_func.h and wrap it in an #ifdef CONFIG_ARMV7_R check or leave it here.  I don't have a strong preference although I tend to use the path that avoids more #if's.

bbolen commented on arch/arm/core/irq_manage.c:
```diff
@@ -16,7 +16,12 @@
 #include <arch/cpu.h>
+#ifdef CONFIG_CPU_CORTEX_M
 #include <arch/arm/cortex_m/cmsis.h>
+#else
```
I'm sorry, I think I'm missing something here.  I was assuming that this code applied to all Cortex-M's.  Are you saying that it should only apply to the M3?

bbolen commented on arch/arm/core/isr_wrapper.S:
```diff
@@ -41,6 +41,22 @@ GTEXT(_IntExit)
+	sub r4, lr, #4
+
+	cps #MODE_SYS
+	stmdb sp!, {r0-r4}
```
I'm saving away r12 with stm r0, {r3-r12,sp,lr}^ in __pendsv.  I believed that was sufficient since no one touches r12 from the entry to __svc until the aforementioned instruction.

pizi-nordic replied:
Unfortunately, you are calling some C code before that (for example the read_timer_start_of_isr below), which is allowed to destroy R12 contents.

Similar case exist also in other places. For example `__svc` calls _irq_do_offload().

bbolen replied:
Ok, I understand what you meant now.  I update the PR to include a fix for that.

bbolen commented on arch/arm/core/cortex_r/vector_table.S:
```diff
@@ -0,0 +1,28 @@
+	ldr	pc,	=__svc               /* svc               offset 8 */
+	ldr	pc,	=__prefetch_abort    /* prefetch abort offset  0xc */
+	ldr	pc,	=__data_abort        /* data abort     offset 0x10 */
+	nop                          /*                offset 0x14 */
```
My particular application does not need FIQ and I was planning on not implementing it for now.

Commit:
drivers: interrupts: Add a set_priority callback

Provide a path for irq controller drivers to change properties of an
individual irq using priority and flags fields that come from the device
tree.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
@pizi-nordic @ioannisg I've updated this PR with changes to address your comments.

ioannisg commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	help
+	  This option signifies the use of an ARMv7-R processor
+	  implementation, or the use of an ARMv8-R processor
+	  implementation supporting the Main Extension.
```
Please, clarify: does the -R- architecture have Base and Mainline versions for v8? If not, the help text needs update

ioannisg replied:
Btw, I searched a bit and I didn't see clear indication of Baseline, Mainline terminology outside the -M- architecture...

ioannisg commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	  referred to as a Mainline Implementation.
+	  - ARMv7-R compatibility requires the Main Extension.
+
+	  From https://developer.arm.com/products/architecture/m-profile:
```
Link points to -M- documentation

ioannisg commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	depends on ARMV7_R
+	help
+	  This option signifies the use of an ARMv7-R processor
+	  implementation, or the use of an ARMv8-R processor
```
ARMv8 is not reflected in the symbol name.

ioannisg commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool
+	select CPU_CORTEX_R
+	select ARMV7_R
+	select ARMV7_M_ARMV8_M_FP if CPU_HAS_FPU
```
ARMV7_M_ARMV8_M_FP  is for Cortex-M CPUs

bbolen replied:
I'm sorry.  Most of this is just copy-paste crud that I'll get rid of.

ioannisg commented on arch/arm/core/cpu_idle.S:
```diff
@@ -120,7 +124,8 @@ SECTION_FUNC(TEXT, k_cpu_idle)
 #endif
 
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
+#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) \
```
This, here (and right below on line 180/185)  look a bit weird; ARMV7_R is "grouped" together with ARMV6_M, instead of ARMV7_M. 

If this is indeed correct, and needs to be as is, it will confuse the reader.
Any suggestion how to address this, @galak ?

ioannisg commented on arch/arm/core/exc_exit.S:
```diff
@@ -69,13 +70,16 @@ SECTION_SUBSEC_FUNC(TEXT, _HandlerModeExit, _IntExit)
     ldr r1, [r0, #_kernel_offset_to_current]
 
-    ldr r0, [r0, _kernel_offset_to_ready_q_cache]
+    ldr r0, [r0, #_kernel_offset_to_ready_q_cache]
```
What was the reason for this?

bbolen replied:
Using the zephyr-sdk, I get a compile error
src/zephyr/arch/arm/core/exc_exit.S: Assembler messages:
src/zephyr/arch/arm/core/exc_exit.S:82: Error: immediate expression requires a # prefix -- `ldr r0,[r0,(0x1c+0x0)]' since _kernel_offset_to_ready_q_cache is an immediate value.  I can't really say why the M boards do not see this.

ioannisg commented on arch/arm/core/swap_helper.S:
```diff
@@ -60,6 +60,7 @@ SECTION_FUNC(TEXT, __pendsv)
 #elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
     movs.n r0, #_EXC_IRQ_DEFAULT_PRIO
     msr BASEPRI, r0
+#elif defined(CONFIG_ARMV7_R)
```
intentionally empty?
please add an explanatory comment

ioannisg commented on arch/arm/core/swap_helper.S:
```diff
@@ -140,7 +151,7 @@ SECTION_FUNC(TEXT, __pendsv)
 #else
     ldr r0, [r2, #_thread_offset_to_basepri]
-    movs.n r3, #0
+    movs r3, #0
```
what is the reason for this change?

bbolen replied:
Again, compile error when using the SDK for Cortex-R
src/zephyr/arch/arm/core/swap_helper.S: Assembler messages:
src/zephyr/arch/arm/core/swap_helper.S:158: Error: unexpected character `n' in type specifier
src/zephyr/arch/arm/core/swap_helper.S:158: Error: bad instruction `movs.n r3,#0'

ioannisg replied:
@galak ^^

ioannisg reviewed (changes requested):
Some more suggestions and issues/questions.

One general thing: Please, use #if defined (), instead of #ifdef in the arm/core and include/ sub-trees

bbolen commented:
Updated again for all the merge conflicts that keep coming in.

@ioannisg What other changes do I need to make in order for this PR to get merged?

galak commented:
@wjliang Can you look at adding support for Xilinix SoC & Ultra96 board

timork commented:
What a status of this PR?

galak commented on include/dt-bindings/interrupt-controller/arm-gic.h:
```diff
@@ -0,0 +1,21 @@
+#define __DT_BINDING_ARM_GIC_H
+
+/* CPU Interrupt numbers */
+#define	GIC_INT_VIRT_MAINT		25
```
These values seem like the would be platform/SoC specific?

bbolen replied:
They are called out in the general GIC400 datasheet, http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0471b/index.html section 2.3.2...

wjliang commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	default n
+
+config XIP
+	default y
```
This is board specific configuration, better to remove this default setting as some platform may not have XIP flash

bbolen replied:
I'm not sure it is completely board specific.  The SoC design can drive whether XIP is necessary or not, e.g. most of the Cortex M implementations.  I'll change it though, as I don't particularly care either way on this one.

Commit:
arch: arm: Add Cortex-R5 support

Pass the correct -mcpu flags to the compiler when building for the
Cortex-R5.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
timer: Add Xilinx ZynqMP PS ttc timer

Add Xilinx PS ttc timer for Xilinx ZynqMP platform.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>

Commit:
serial: Add Xilinx ZynqMP PS uart driver

Add ZynqMP PS uart driver for Xilinx ZynqMP platform

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>

wjliang commented on arch/arm/core/irq_manage.c:
```diff
@@ -115,6 +121,74 @@ void _irq_priority_set(unsigned int irq, unsigned int prio, u32_t flags)
+{
+	struct device *dev = _sw_isr_table[0].arg;
+
+	irq_set_priority_next_level(dev, (irq >> 8) - 1, prio, flags);
```
could you clarify on why it needs "(irq >> 8) -1"?  how about just irq? in our case, we have SPI, which starts from 32, i am not clear on why the code look like it reserve the irq 0 - 256, and let each platform design what irq should be passed in from device tree?

bbolen replied:
In order to reuse isr_wrapper.S when the irq exception is hit, I made the assumption that the GIC (or VIC) will be the first level interrupt controller, so offset 0 of the isr table should have the device and callback of the gic.  The interrupts coming into the GIC are then 2nd level interrupts which use the second nibble of the irq number.  It's not that 0-255 are reserved, it's that the decision on the nesting hierarchy of the interrupt controllers forces that model where the individual interrupt numbers are in the second nibble.

wjliang replied:
Does it mean we assumes that 0xFF is the 1st level? and is it always the case in Zephyr? and thus, for cortex-R5, all the interrupts mask will be 0xFFFFFF00?
Just as R5 doesn't use 1nd level at all, can we  assume  all the interrupt is for 2nd level, and let the conversion to be done inside the GIC controller?

bbolen replied:
The least significant two nibbles will be the first level.  That's always the case for Zephyr.  For the Cortex R cpus, the difference is that the hardware exception table only has one slot for the interrupt handler.  The Cortex R itself does not handle figuring out what peripheral generated the interrupt (unlike Cortex M).  

The Cortex R5 does use 1st level interrupts, it's just that there is only one first level interrupt handler, the GIC.

Note that the GIC is not the only interrupt controller that could be hooked in here.  In my specific case, I am now using an arm pl192 (vic), so I don't think the interrupt controller driver should know anything about the "level" of its irqs.  The number passed into the irq handler should be meaningful only in the frame of reference of the irq handler, not the entire system.  For example, if you wanted to call irq_enable(0x1200) for irq 18 on the gic, my opinion is that the number coming in the parameter to gic_irq_enable should be 18, not 0x1200.

Ideally, the device tree parser should be able to generate the irq tree and handle shifting irq numbers based on where the irq is in the tree, without needing to fixup irq numbers.  But the work is beyond the scope of my knowledge right now.

I did play with trying to put the "shifted" values in the device tree itself, but found that the build system produced a huge isr_sw_table that contained mostly holes of unused irq entries.

wjliang replied:
I have  tried your suggestion to keep the origin value of interrupt in DTS but use the dts fix up to shift the value. It worked for me.

wjliang commented on drivers/interrupt_controller/CMakeLists.txt:
```diff
@@ -10,3 +10,4 @@ zephyr_sources_ifdef(CONFIG_SOC_FAMILY_STM32        exti_stm32.c)
 zephyr_sources_ifdef(CONFIG_CAVS_ICTL               cavs_ictl.c)
 zephyr_sources_ifdef(CONFIG_DW_ICTL                 dw_ictl.c)
 zephyr_sources_ifdef(CONFIG_RV32M1_INTMUX           rv32m1_intmux.c)
+zephyr_sources_ifdef(CONFIG_GIC                     gic.c)
```
just minor comments, how about the name change to gic-400 to match the implementation?

bbolen replied:
done

wjliang commented on dts/arm/armv7-r.dtsi:
```diff
@@ -0,0 +1,34 @@
+			reg = <0x00 0x4>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
```
This DTS looks platform dependent.

bbolen replied:
I replaced the reference to "R4" with "Cortex-R", but I'm not sure what else would be considered platform dependent...

wjliang commented:
This pull request has been tested on  Xilinx Zynq UltraScale+ MP SoC Cortex-R5 processor.

Commit:
Revert "toolchain: gcc: Remove ref. to undefined sym. CONFIG_ISA_ARM"

This reverts commit e20fd5f454b4b4e3ed7b94d88ac514c3f9928c04.

This is needed for Cortex-R support.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Add memory barriers for register accesses

Cortex R has a write buffer that can cause reordering problems when
accessing memory mapped registers.  Use memory barries to make sure that
these accesses are performed in the desired order.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Mani-Sadhasivam commented:
@wjliang Will you be contributing the soc support you have tested? Otherwise, I may have to do it along with Ultra96 board support.

Mani-Sadhasivam commented:
@bbolen I'm working on adding Xilinx Ultrascale+ MPSoC and  Ultra96 board support based on this PR. Could you please share your custom soc/board source which you are using for testing as a reference?

Mani-Sadhasivam commented:
@bbolen NVM, I got it from your tree ;-)
https://github.com/bbolen/zephyr/commit/068405f9de6c108be497db86dea60cbf34f77f5c

bbolen commented:
@galak @nashif Is there any timeline on when/if this PR will be merged?  Keeping up with the ongoing merge conflicts is becoming frustrating.  If it is going to be a while, then I may let the merge conflicts sit for a while.

carlescufi commented:
> @galak @nashif Is there any timeline on when/if this PR will be merged? Keeping up with the ongoing merge conflicts is becoming frustrating. If it is going to be a while, then I may let the merge conflicts sit for a while.

@bbolen I think there is quite a few outstanding review comments you haven't replied to. 
Also as per conversations with @galak and @nashif, we really need a board to integrate this into CI. Ideally qemu, is that something you could add to this PR? We typically don't allow new functionality that cannot be tested on CI

bbolen commented:
@carlescufi  github has collapsed some of the history on this so you may need to go expand it to see the entire history.

Re review comments
From the history above:
"bbolen commented on Oct 24, 2018
I made a couple of tweaks so that Cortex-R uses multilevel  IRQs.  The first level is always the isr_wrapper.  The second level is  the system interrupt controller (GIC, PL192 VIC, etc).  The code as it  was would not work if there was an interrupt wired to input 0 of the  system interrupt controller. @ioannisg I should have addressed all of the previous comments."
--

I did not reply to each individual comment, but responded to the whole thing in Oct 2018.  Do I need to go hit the "Resolve Conversation" button on each of those?  I have handled every comment and have done a years worth of rebasing on this PR to keep it up to date.  I have not seen any other review comments since I responded.

Re board requirements
From the history above:
" wjliang commented on Feb 18
This pull request has been tested on Xilinx Zynq UltraScale+ MP SoC Cortex-R5 processor."
@galak and @wjliang were working on board support for this and had something working which I assumed would get lumped with the PR at some point.  But I have heard nothing from either of them since Feb.

If no one wants this then I'll just close it, but I've seen multiple examples of people using this code and some wanting to adapt it to Cortex-A as well.  I just cannot get any traction to know what else is needed to push this to completion since I only got sporadic responses over the past year.

carlescufi commented:
@bbolen thanks for taking the time to respond in detail.

>I did not reply to each individual comment, but responded to the whole thing in Oct 2018. Do I need to go hit the "Resolve Conversation" button on each of those? I have handled every comment and have done a years worth of rebasing on this PR to keep it up to date. I have not seen any other review comments since I responded.

No, if you addressed or responded to the comments (I was mainly referring to those posted by @pizi-nordic and @ioannisg), then there is no need to do anything else, but then they should either approve or request for additional changes.

>If no one wants this then I'll just close it, but I've seen multiple examples of people using this code and some wanting to adapt it to Cortex-A as well. I just cannot get any traction to know what else is needed to push this to completion since I only got sporadic responses over the past year.

I do not think there is any debate about whether we want this upstream: we do. But if we do merge it without an SoC and board, it will likely bitrot in the tree, because no one will notice if unrelated changes break it. I've spoken to @galak and unfortunately the Xilinx work you was promised is not progressing as fast as he'd like. Is there any chance you could add a very simple SoC/board based on QEMU? Because that would be ideal: we'd be able to build samples on this new arch, and you would not need to wait for a port. If that is not possible, how about adding your own SoC/board, the one you are actually  using to test this? Would that be an option?

Sorry that it has taken so long and that you've had to keep this rebased multiple times.

carlescufi commented:
@ioannisg can you please take another look at the code and approve if appropriate?

bbolen commented:
@carlescufi I'll take a look at seeing if I can add SoC code based on the xlnx-zynqmp qemu platform.  What minimal set of sample applications should be able to run for the CI test framework?

carlescufi commented:
> @carlescufi I'll take a look at seeing if I can add SoC code based on the xlnx-zynqmp qemu platform. What minimal set of sample applications should be able to run for the CI test framework?

If you manage to build the `hello_world` sample and run it on qemu we will already be well on the way there. If you had a UART driver it would even be better because then we'd be able to run the kernel tests on qemu, which is really the most important requirement.

Commit:
soc: arm: xilinx_zynqmp: Add qemu based SoC

This commit adds support for the Zynq UltraScale+ MPSoC as a qemu based
platform for Cortex-R based testing.  This SoC only supports an
interrupt controller and serial port for limited testing.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
boards: qemu_cortex_r5: Add qemu test board for the Cortex-R series

This adds a qemu test board using the Xilinx ZynqMP SoC.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
> > @carlescufi I'll take a look at seeing if I can add SoC code based on the xlnx-zynqmp qemu platform. What minimal set of sample applications should be able to run for the CI test framework?
> 
> If you manage to build the `hello_world` sample and run it on qemu we will already be well on the way there. If you had a UART driver it would even be better because then we'd be able to run the kernel tests on qemu, which is really the most important requirement.

I was able to get the hello world sample running on the xlnx-zcu102 qemu machine.  However, that platform uses the qemu-system-aarch64 binary which isn't included in the zephyr sdk.  I don't know which path is appropriate, add aarch64 qemu support to the zephyr-sdk, or gut the xlnx-zcu102 qemu machine to remove the A53's so that it can be run with the 32 bit qemu binary?

andrewboie commented:
>  add aarch64 qemu support to the zephyr-sdk, 

This is what we should do, I think.
@nashif is on vacation for the next 3 weeks, @galak @carlescufi @MaureenHelm who handles SDK stuff other than Anas?

carlescufi commented:
> > add aarch64 qemu support to the zephyr-sdk,
> 
> This is what we should do, I think.
> @nashif is on vacation for the next 3 weeks, @galak @carlescufi @MaureenHelm who handles SDK stuff other than Anas?

@galak does. @bbolen please contact @galak directly or on [Slack](https://tinyurl.com/yarkuemx)

Commit:
cmake: emu: Allow overriding the qemu binary

The xlnx-zcu102 qemu machine is the only one that supports a Cortex-R
processor.  However, its main CPUs are Cortex A53s which requires the
aarch64 qemu binary to run.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

andrewboie commented on arch/arm/core/cortex_r/fault.c:
```diff
@@ -0,0 +1,28 @@
+ */
+void _Fault(z_arch_esf_t *esf, u32_t exc_return)
+{
+	z_arm_fatal_error(K_ERR_KERNEL_PANIC, NULL);
```
please pass along the esf pointer instead of NULL.
This should be K_ERR_CPU_EXCEPTION, not K_ERR_KERNEL_PANIC

bbolen replied:
fixed

andrewboie commented on arch/arm/core/cortex_r/fault.c:
```diff
@@ -0,0 +1,32 @@
+/*
```
I don't think that's necessary since it is obvious what is going on.
More fleshed out exception handling can take place in this C file.

andrewboie commented on arch/arm/core/cortex_r/irq_init.c:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @return N/A
+ */
+
```
Does this really need to be in its own C file?
Please just declare as a static inline function in kernel_arch_func.h

bbolen replied:
fixed

andrewboie commented on include/arch/arm/error.h:
```diff
@@ -53,6 +53,11 @@ extern "C" {
 } while (false)
+#elif defined(CONFIG_ARMV7_R)
+#define Z_ARCH_EXCEPT(reason_p) do { \
+	k_sys_fatal_error_handler(K_ERR_KERNEL_PANIC, NULL); \
```
you should not be invoking k_sys_fatal_error_handler() directly! for one thing, it can return.

either leave this undefined so that the default definition in include/kernel.h is used, or better yet get the kernel oops code in the SVC handler working on this arch

bbolen replied:
I let it pick up the default definition for now.  I will fix it to use the svc handler in a future PR.

tyrantlizard77 reviewed (changes requested):
Overall looks good just some complaints related to fatal errors.

ioannisg commented:
@bbolen (and all)
Sorry for not attending this PR for such long time. I'll have some time to review this now.

I would really appreciate if you could do the following; split this big PR in smaller parts. The first part would be **_only_** (but also **_all_**) the refactoring (e.g. moving arch/arm header and source files, etc.) to make the code base friendly, so as to add the Cortex-R support right after. 

_That_ smaller patch will be very easy to review, revise and merge that. As soon as we merge, you can rebase the bigger patch with the actual Cortex-R support.

What do you think about that @bbolen @galak @carlescufi ?

Commit:
arch: arm: Add Cortex-R4 support

Pass the correct -march and -mcpu flags to the compiler when building
for the Cortex-R4.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

galak commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.dts:
```diff
@@ -0,0 +1,43 @@
+
+	flash0: flash@c0000000 {
+		compatible = "soc-nv-flash";
+		reg = <0xc0000000 (64 * 1024)>;
```
This can be DT_SIZE_K(64)

bbolen replied:
fixed

galak commented on dts/bindings/timer/xlnx,ttcps.yaml:
```diff
@@ -0,0 +1,24 @@
+    reg:
+      category: required
+
+    clock-frequency:
```
Seems like this should be clock instead.

bbolen replied:
Are you meaning to create to clock_control instance for the lpd_lsbus_clk that feeds the timers?

ioannisg commented on CODEOWNERS:
```diff
@@ -153,6 +153,8 @@
 /drivers/serial/uart_liteuart.c           @mateusz-holenko @kgugala @pgielda
 /drivers/serial/Kconfig.rtt               @carlescufi @pkral78
 /drivers/serial/uart_rtt.c                @carlescufi @pkral78
+/drivers/serial/Kconfig.xlnx              @wjliang
```
Btw, @bbolen would you like to add someone in the list of maintainers for the arch/arm/core/cortex-r ?

bbolen replied:
Added myself.

ioannisg commented on arch/arm/core/CMakeLists.txt:
```diff
@@ -33,3 +33,5 @@ add_subdirectory_ifdef(CONFIG_ARM_MPU cortex_m/mpu)
 add_subdirectory_ifdef(CONFIG_ARM_SECURE_FIRMWARE cortex_m/tz)
 add_subdirectory_ifdef(CONFIG_ARM_NONSECURE_FIRMWARE cortex_m/tz)
+
+add_subdirectory_ifdef(CONFIG_CPU_CORTEX_R cortex_r)
```
There seems to be an issue here, at least, with ARM_MPU, which brings-in cortex-m/mpu, but the ARM_MPU is also implied to be used in CORTEX_R (see config CPU_CORTEX_R)

ioannisg commented on arch/arm/core/Kconfig:
```diff
@@ -29,6 +29,98 @@ config CPU_CORTEX_M
+	  This is needed to conform to AAPCS, the procedure call standard for
+	  the ARM. It wastes stack space. The option also enforces alignment
+	  of stack upon exception entry on Cortex-M3 and Cortex-M4 (ARMv7-M).
+	  Note that for ARMv6-M, ARMv8-M, and Cortex-M7 MCUs stack alignment
```
Any particular comments needed for Cortex-R here?

bbolen replied:
Shouldn't need anything.  The AAPCS comment covers the main reason and this config option isn't used on cortex-r to effect the exception stacks.

ioannisg replied:
OK

ioannisg commented on arch/arm/core/exc_exit.S:
```diff
@@ -23,6 +23,7 @@ _ASM_FILE_PROLOGUE
 
 GTEXT(z_ExcExit)
 GTEXT(_IntExit)
+GTEXT(__pendsv)
```
Should be in #ifdef Cortex-R guards

bbolen replied:
fixed

ioannisg commented on arch/arm/core/fault_s.S:
```diff
@@ -127,6 +135,7 @@ _s_stack_frame_msp:
 	mrs r0, MSP
 _s_stack_frame_endif:
 #endif /* CONFIG_ARM_SECURE_FIRMWARE || CONFIG_ARM_NONSECURE_FIRMWARE */
+#elif defined(CONFIG_ARMV7_R)
```
MAybe an inline comment justifying why this is empty?

bbolen replied:
fixed

ioannisg replied:
ok

ioannisg commented on arch/arm/core/Kconfig:
```diff
@@ -29,6 +29,98 @@ config CPU_CORTEX_M
+	select CPU_CORTEX
+	select HAS_FLASH_LOAD_OFFSET
+	select ARCH_HAS_STACK_PROTECTION if ARM_MPU
+	select ARCH_HAS_USERSPACE if ARM_MPU
```
Here, you imply that ARM_MPU may be selected for Cortex-R, but in CMakeLists, CONFIG_ARM_MPU brings-in cortex_m/mpu folder

bbolen replied:
I removed these two lines.  Just a heads-up that we are working on MPU support for Cortex-R so that code will be submitted at some point.

ioannisg commented on arch/arm/core/cpu_idle.S:
```diff
@@ -21,19 +21,17 @@
 _ASM_FILE_PROLOGUE
 
 GTEXT(z_CpuIdleInit)
-#ifdef CONFIG_SYS_POWER_MANAGEMENT
```
Could you please, remind me why this has been cleaned-up?

ioannisg replied:
OH, I see, it is arch: arm: cpu_idle: Remove unused functions 
Could you split that out in a standalone PR? That's a clean up patch - not belonging here.

ioannisg replied:
..and that PR can be reviewed very fast, of course.

bbolen replied:
Submitted #18099

ioannisg commented:
Pls, rebase once more @bbolen

ioannisg commented on include/arch/arm/syscall.h:
```diff
@@ -16,6 +16,7 @@
 #ifndef ZEPHYR_INCLUDE_ARCH_ARM_SYSCALL_H_
 #define ZEPHYR_INCLUDE_ARCH_ARM_SYSCALL_H_
 
+#define _SVC_CALL_CONTEXT_SWITCH	0
```
This value was not reserved for anything else, right @andrewboie ?

bbolen commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.dts:
```diff
@@ -0,0 +1,43 @@
+	sram0: memory@0 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0 (256 * 1024)>;
```
fixed

ioannisg commented on CODEOWNERS:
```diff
@@ -18,6 +18,7 @@
 /arch/arm/                                @MaureenHelm @galak @ioannisg
 /arch/arm/core/cortex_m/cmse/             @ioannisg
 /arch/arm/include/cortex_m/cmse.h         @ioannisg
+/arch/arm/core/cortex_r/                  @bbolen
```
Last entry get precedence, so please, add the above ARM maintainers in this line.

bbolen replied:
fixed

ioannisg reviewed (commented):
Some comments.
The /arch/arm changes look OK, imo.
I am going to go through the rest of the review later today

ioannisg commented on arch/arm/core/isr_wrapper.S:
```diff
@@ -128,12 +160,17 @@ _idle_state_cleared:
 	mov lr, r3
 #elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
 	pop {r0, lr}
+#elif defined(CONFIG_ARMV7_R)
```
This #elif is empty, so maybe write something there..

bbolen replied:
added comment

ioannisg commented on arch/arm/core/isr_wrapper.S:
```diff
@@ -128,12 +160,17 @@ _idle_state_cleared:
 	 * _IntExit() */
-	ldr r0, =_IntExit
-	bx r0
+	ldr r1, =_IntExit
```
Any reason this was changed @bbolen ?

bbolen replied:
Exiting from an interrupt and exiting from a svc call require slightly different paths in arch/arm/core/exc_exit.S:z_ExcExit().  I'm using r0 to pass which context it is leaving from in z_ExcExit().  I could have left that code alone and just passed the return context in r1, but that seemed a little weird having an argument coming into z_ExcExit in r1, but nothing in r0.

ioannisg commented on arch/arm/core/prep_c.c:
```diff
@@ -150,9 +155,6 @@ extern FUNC_NORETURN void z_cstart(void);
  * @return N/A
  */
-
-extern void z_IntLibInit(void);
```
Is this included, now, via some header?

ioannisg commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.yaml:
```diff
@@ -0,0 +1,48 @@
+flash: 256
+testing:
+  default: false
+  ignore_tags:
```
Not sure why these tags are all ignored

bbolen replied:
Currently, the cortex R5 on the xlnx-zynmp qemu machine is not hooked up to an interrupt controller, so only the "hello world" test case will run right now.

ioannisg commented on drivers/serial/Kconfig.xlnx:
```diff
@@ -0,0 +1,14 @@
+	select SERIAL_HAS_DRIVER
+	help
+	  This option enables the UART driver for Xilinx MPSoC platforms.
+	  Say y if you wish to use serial port on Xilinx MPSoC platforms.
```
Not sure if you really need the last line

bbolen replied:
fixed

ioannisg reviewed (approved):
Went through the PR, looks OK to me. I still left several comments to address, but, just to try to push this for 2.0, I give a +1

ioannisg commented:
@bbolen I suppose if you rebase once more you can run CI with the new SDK, Qemu.

bbolen commented on include/arch/arm/irq.h:
```diff
@@ -34,6 +34,14 @@ extern int z_arch_irq_is_enabled(unsigned int irq);
+{
+}
+#else
+extern void z_IntLibInit(void);
```
Yes, I originally had an empty definition for z_IntLibInit in its own file.  andrewbowie asked me to move it to a header file, so it made more sense to move the local declaration out of prep_c.c as well.

ioannisg commented:
@andrewboie @galak could you have another look? 
@bbolen could you rebase to latest master?
